core/vm/gas_table.go

```diff
@@ -278,24 +278,59 @@ var (
 	gasMLoad   = pureMemoryGascost
 	gasMStore8 = pureMemoryGascost
 	gasMStore  = pureMemoryGascost
-	gasCreate  = pureMemoryGascost
 )
 
+func gasCreate(evm *EVM, contract *Contract, stack *Stack, mem *Memory, memorySize uint64) (uint64, error) {
+	gas, err := memoryGasCost(mem, memorySize)
+	if err != nil {
+		return 0, err
+	}
+	len, overflow := stack.Back(2).Uint64WithOverflow()
+	if overflow {
+		return 0, ErrGasUintOverflow
+	}
+
+	if hasEIP3860(&evm.Config) {
+		lenWords := toWordSize(len)
+		initCodeWordGas, overflow := math.SafeMul(lenWords, params.InitCodeWordGas)
+		if overflow {
+			return 0, ErrGasUintOverflow
+		}
+		if gas, overflow = math.SafeAdd(gas, initCodeWordGas); overflow {
+			return 0, ErrGasUintOverflow
+		}
+	}
+	return gas, nil
+}
+
 func gasCreate2(evm *EVM, contract *Contract, stack *Stack, mem *Memory, memorySize uint64) (uint64, error) {
 	gas, err := memoryGasCost(mem, memorySize)
 	if err != nil {
 		return 0, err
 	}
-	wordGas, overflow := stack.Back(2).Uint64WithOverflow()
+	len, overflow := stack.Back(2).Uint64WithOverflow()
 	if overflow {
 		return 0, ErrGasUintOverflow
 	}
-	if wordGas, overflow = math.SafeMul(toWordSize(wordGas), params.Sha3WordGas); overflow {
+	lenWords := toWordSize(len)
+
+	hashingWordGas, overflow := math.SafeMul(lenWords, params.Sha3WordGas)
```

Comment:
If this is moved to be the last step, then the size-related check can be deduplicated between `gasCreate` and `gasCreate2`. I think we also mentioned this in the EIP.

Not sure if this is a concern in geth, but thought to mention it.

Reply:
It would be a bit tricky without having to repeat `len, overflow := stack.Back(2).Uint64WithOverflow()` and `lenWords := toWordSize(len)` calls, so I decided that maybe it's simpler to leave duplication.